smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -532,20 +569,326 @@ private void generateOperationDeserializer(
+            return "new Date(" + dataSource + ")";
+        } else if (target instanceof BlobShape) {
+            return getBlobOutputParam(bindingType, dataSource);
+        } else if (target instanceof CollectionShape) {
+            return getCollectionOutputParam(context, bindingType, dataSource, target);
+        } else if (target instanceof StructureShape || target instanceof UnionShape || target instanceof MapShape) {
+            Symbol symbol = symbolProvider.toSymbol(target);
+            return ProtocolGenerator.getDeserFunctionName(symbol, getName()) + "(" + dataSource + ", context)";
+        }
+
+        throw new CodegenException(String.format(
+                "Unsupported %s binding of %s to %s in %s using the %s protocol",
+                bindingType, dataSource, target.getType(), container, getName()));
+    }
+
+    private String getBlobOutputParam(Location bindingType, String dataSource) {
+        switch (bindingType) {
+            case PAYLOAD:
+                return dataSource;
+            case HEADER:
+            case DOCUMENT:
+                // Decode these from base64.
+                return "context.base64Decoder.fromBase64(" + dataSource + ")";
+            default:
+                throw new CodegenException("Unexpected blob binding location`" + bindingType);
+        }
+    }
+
+    private String getCollectionOutputParam(
+            GenerationContext context,
+            Location bindingType,
+            String dataSource,
+            Shape target
+    ) {
+        switch (bindingType) {
+            case HEADER:
+                // TODO Is special handling needed for string contents with commas and/or newlines?
+                // Split these values on commas.
+                return dataSource + ".split(',')";
+            case DOCUMENT:
+                SymbolProvider symbolProvider = context.getSymbolProvider();
+                Symbol symbol = symbolProvider.toSymbol(target);
+
+                return ProtocolGenerator.getDeserFunctionName(symbol, getName()) + "(" + dataSource +  ", context)";
+            default:
+                throw new CodegenException("Unexpected collection binding location`" + bindingType);
+        }
+    }
+
+    protected boolean unionTargetNeedsSerde(ShapeIndex index, Shape shape) {
+        if (shape instanceof SimpleShape) {
+            return false;
+        }
+
+        return (!(shape instanceof CollectionShape) || collectionNeedsSerde(index, (CollectionShape) shape));
+    }
+
+    protected boolean collectionNeedsSerde(ShapeIndex index, CollectionShape shape) {
+        Shape target = index.getShape(shape.getMember().getTarget()).get();
+        if (target instanceof CollectionShape) {
+            return collectionNeedsSerde(index, (CollectionShape) target);
+        }
+        return !(target instanceof SimpleShape);
+    }
+
+    private final class ShapeSerializingVisitor extends ShapeVisitor.Default<Void> {
```

Comment:
It's a lot to ask... but examples of what each branch generates would be suuuper helpful